src/resmom/linux/alps.c

```diff
@@ -6397,129 +6665,916 @@ alps_inventory(void)
+ *
+ * @param[in] sys_knl ALPS BASIL System Query response.
+ *
+ * @return void
+ */
+static void
+create_vnodes_KNL(basil_response_query_system_t *sys_knl)
+{
+	char	*attr, *arch;
+	char	vname[VNODE_NAME_LEN];
+	char	utilBuffer_knl[(UTIL_BUFFER_LEN * sizeof(char))];
+	int	ncpus_per_knl;
+	int	node_idx = 0;
+	int	node_count =0;
+	int	node_id = 0;
+	int	*nid_arr = NULL;
+	int	atype = READ_WRITE | ATR_DFLAG_CVTSLT;
+	char	mpphost_knl[BASIL_STRING_LONG];
+
+
+	basil_system_element_t *node_group;
+
+
+	if (sys_knl == NULL)
+		return;
+
+	strncpy(mpphost_knl, sys_knl->mpp_host, sizeof(mpphost_knl) - 1);
+	mpphost_knl[sizeof(mpphost_knl) - 1] = '\0';
+
+
+	/*
+	 * Iterate through all the Node groups in the System Query Response. Each
+	 * Node group may contain information about multiple (a 'range list') of KNL Nodes.
+	 * Each XML Element <Nodes ...> </Nodes> encapsulates information about a group of Nodes.
+	 */
+	for (node_group = sys_knl->elements; node_group; node_group = node_group->next) {
+
+		/*
+		 * The System Query XML Response contains information about KNL and
+		 * non-KNL Nodes. We are only interested in KNL nodes that are in
+		 * "batch" mode and in the "up" state.
+		 */
+		if (exclude_from_KNL_processing(node_group))
+			continue;
+
+		/*
+		 * Extract NIDs (Node ID) from node_group->nidlist.
+		 * If nidlist is empty, node_count gets set to 0 and vnode
+		 * creation in the inner for() is bypassed.
+		 */
+		nid_arr = process_nodelist_KNL(node_group->nidlist, &node_count);
+
+		/*
+		 * Create vnodes for each of the KNL Nodes listed in this Node group.
+		 * All KNL Nodes within a Node Group will have similar vnode attributes,
+		 * since all attributes in each <Nodes ...> XML element apply to all
+		 * Nodes listed in the 'rangelist'.
+		 */
+		for (node_idx = 0; node_idx < node_count; node_idx++) {
+
+			node_id = nid_arr[node_idx];
+			snprintf(vname, VNODE_NAME_LEN, "%s_%d", mpphost_knl, node_id);
+
+			if (first_compute_node) {
+				/*
+				 * Create the name of the very first vnode so we
+				 * can attach topology info to it.
+				 */
+
+				attr = ATTR_NODE_TopologyInfo;
+				if (vn_addvnr(vnlp, vname, attr, (char *) basil_inventory,
+					     ATR_TYPE_STR, READ_ONLY, NULL) == -1)
+					  goto bad_vnl;
+				first_compute_node = 0;
+			}
+
+			attr = "sharing";
+			if (vn_addvnr(vnlp, vname, attr, ND_Force_Exclhost, 0, 0, NULL) == -1)
+				goto bad_vnl;
+
+			attr = "resources_available.vntype";
+			if (vn_addvnr(vnlp, vname, attr, CRAY_COMPUTE, 0, 0, NULL) == -1)
+				goto bad_vnl;
+
+			attr = "resources_available.PBScrayhost";
+			if (vn_addvnr(vnlp, vname, attr, mpphost_knl, ATR_TYPE_STR, atype, NULL) == -1)
+				goto bad_vnl;
+
+			attr = "resources_available.arch";
+			arch = BASIL_VAL_XT;
+			if (vn_addvnr(vnlp, vname, attr, arch, ATR_TYPE_STR, atype, NULL) == -1)
+				goto bad_vnl;
+
+			attr = "resources_available.host";
+			snprintf(utilBuffer_knl, sizeof(utilBuffer_knl), "%s_%d", mpphost_knl, node_id);
```

Comment:
Once the node_id being returned is a long, then this should be updated to handle a long.